test/helper/log_and_filtering.bash

```diff
@@ -0,0 +1,109 @@
+# ## Attention
+#
+# The string given to this function is interpreted by `grep -E`, i.e.
+# as a regular expression. In case you use characters that are special
+# in regular expressions, you need to escape them!
+function _filter_service_log() {
+  local SERVICE=${1:?Service name must be provided}
+  local STRING=${2:?String to match must be provided}
+  local CONTAINER_NAME=$(__handle_container_name "${3:-}")
+  local FILE="/var/log/supervisor/${SERVICE}.log"
+
+  # Fallback to alternative log location:
+  [[ -f ${FILE} ]] || FILE="/var/log/mail/${SERVICE}.log"
+  _run_in_container grep -i -E "${STRING}" "${FILE}"
+}
+
+# Like `_filter_service_log` but asserts that the string was found.
+#
+# @param ${1} = service name
+# @param ${2} = string to filter by
+# @param ${3} = container name [OPTIONAL]
+#
+# ## Attention
+#
+# The string given to this function is interpreted by `grep -E`, i.e.
+# as a regular expression. In case you use characters that are special
+# in regular expressions, you need to escape them!
+function _service_log_should_contain_string() {
+  _filter_service_log "${@}"
+  assert_success
+}
+
+# Like `_filter_service_log` but asserts that the string was not found.
+#
+# @param ${1} = service name
+# @param ${2} = string to filter by
+# @param ${3} = container name [OPTIONAL]
+#
+# ## Attention
+#
+# The string given to this function is interpreted by `grep -E`, i.e.
+# as a regular expression. In case you use characters that are special
+# in regular expressions, you need to escape them!
+function _service_log_should_not_contain_string() {
+  _filter_service_log "${@}"
+  assert_failure
+}
+
+# Filters the mail log according to MID (Message-ID) and prints lines
+# of the mail log that fit Postfix's queue ID for the given message ID.
+#
+# @param ${1} = message ID part before '@'
+function _print_mail_log_of_queue_id_from_mid() {
+  # The unique ID Postfix (and other services) use may be different in length
+  # on different systems. Hence, we use a range to safely capture it.
+  local QUEUE_ID_REGEX='[A-Z0-9]{9,12}'
+
+  local MID=$(__construct_mid "${1:?Left-hand side of MID missing}")
+  shift 1
+
+  _wait_for_empty_mail_queue_in_container
+
+  QUEUE_ID=$(_exec_in_container tac /var/log/mail.log                    \
+    | grep -E "postfix/cleanup.*: ${QUEUE_ID_REGEX}:.*message-id=${MID}" \
+    | grep -E --only-matching --max-count 1 "${QUEUE_ID_REGEX}" || :)
+
+  # We perform plausibility checks on the IDs.
+  assert_not_equal "${QUEUE_ID}" ''
+  run echo "${QUEUE_ID}"
+  assert_line --regexp "^${QUEUE_ID_REGEX}$"
+
+  _filter_service_log 'mail' "${QUEUE_ID}"
+}
+
+# Filters the mail log according to MID (Message-ID) and prints lines
+# of the mail log that fit lines with the pattern `msgid=${1}@dms-test`.
+#
+# @param ${1} = message ID part before '@'
+function _print_mail_log_for_msgid() {
+  local MID=$(__construct_mid "${1:?Left-hand side of MID missing}")
+  shift 1
+
+  _filter_service_log 'mail' "msgid=${MID}"
+}
```

Comment:
Postfix uses `message-id`, you may prefer a regex, or to filter specifically for any logs with the `MSG_ID` value as a match, otherwise the current usage is Dovecot specific.

```suggestion
# A convenience method that filters for Dovecot specific logs with a `msgid` field that matches the MSG_ID input.
#
# @param ${1} = The local-part of a Message-ID header value (`<local-part@domain-part>`)
function _print_mail_log_for_msgid() {
  local MSG_ID=$(__construct_msgid "${1:?The local-part for MSG_ID was not provided}")
  shift 1

  _filter_service_log 'mail' "msgid=${MSG_ID}"
}
```